File: C++/labs/lab2/foldingHasher.cpp
// Задание 6 | Вариант 1
#include "foldingHasher.h"

#include <algorithm> 
#include <cstring>
#include <iostream>
#include <sstream> 
#include <string>

long long foldingHash(const std::string& key, int chunkSize) {
    long long totalSum = 0;

    std::cout << "Вычисление: ";
    bool isFirstPart = true;

    // Идем по строке с шагом, равным размеру части
    for (size_t i = 0; i < key.length(); i += chunkSize) {
        // Вырезаем очередную часть из строки
        std::string partStr = key.substr(i, chunkSize);

        // Конвертируем строку в число. Используем stoll для long long.
        long long partValue = std::stoll(partStr);

        // Добавляем к общей сумме
        totalSum += partValue;

        // Выводим процесс вычисления, как в примере
        if (!isFirstPart) {
            std::cout << "+";
        }
        std::cout << partStr;
        isFirstPart = false;
    }

    // Выводим итоговый результат
    std::cout << " = " << totalSum << std::endl;

    return totalSum;
}

void runFoldingHasher(int argc, char* argv[]) {
    std::string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
            break;
        }
    }

    if (query.empty()) {
        std::cerr << "Ошибка: не найден аргумент --query." << std::endl;
        std::cerr << "Пример использования: --query \"FOLD <число>\"" << std::endl;
        return;
    }

    std::stringstream ss(query);
    std::string command, numberKey;
    ss >> command >> numberKey;

    if (command != "FOLD" || numberKey.empty()) {
        std::cerr << "Ошибка: некорректный формат запроса." << std::endl;
        std::cerr << "Пример использования: --query \"FOLD 523456795\"" << std::endl;
        return;
    }

    // Проверка, что ключ состоит только из цифр
    for (char const& c : numberKey) {
        if (std::isdigit(c) == 0) {
            std::cerr << "Ошибка: ключ должен состоять только из цифр." << std::endl;
            return;
        }
    }

    // В примере используется разбиение на части по 3 символа.
    const int chunkSize = 3;

    std::cout << "Ввод: " << numberKey << std::endl;
    std::cout << "Вывод: ";
    foldingHash(numberKey, chunkSize);
}

File: Go/labs/lab2/foldingHasher.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"unicode"
)

func foldingHash(key string, chunkSize int) int64 {
	var totalSum int64
	fmt.Print("Вычисление: ")
	isFirstPart := true

	for i := 0; i < len(key); i += chunkSize {
		end := i + chunkSize
		if end > len(key) {
			end = len(key)
		}
		partStr := key[i:end]

		partValue, _ := strconv.ParseInt(partStr, 10, 64)
		totalSum += partValue

		if !isFirstPart {
			fmt.Print("+")
		}
		fmt.Print(partStr)
		isFirstPart = false
	}
	fmt.Printf(" = %d\n", totalSum)
	return totalSum
}

func runFoldingHasher(args []string) {
	var query string
	for i, arg := range args {
		if arg == "--query" && i+1 < len(args) {
			query = args[i+1]
			break
		}
	}

	if query == "" {
		fmt.Fprintln(os.Stderr, "Ошибка: не найден аргумент --query.")
		fmt.Fprintln(os.Stderr, "Пример использования: --query \"FOLD <число>\"")
		return
	}

	parts := strings.Fields(query)
	if len(parts) != 2 || parts[0] != "FOLD" {
		fmt.Fprintln(os.Stderr, "Ошибка: некорректный формат запроса.")
		fmt.Fprintln(os.Stderr, "Пример использования: --query \"FOLD 523456795\"")
		return
	}
	numberKey := parts[1]

	for _, c := range numberKey {
		if !unicode.IsDigit(c) {
			fmt.Fprintln(os.Stderr, "Ошибка: ключ должен состоять только из цифр.")
			return
		}
	}

	const chunkSize = 3
	fmt.Println("Ввод:", numberKey)
	fmt.Print("Вывод: ")
	foldingHash(numberKey, chunkSize)
}